sql/plan/alter_table.go

```diff
@@ -221,6 +221,9 @@ func (a *AddColumn) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error)
 	if err := a.validateDefaultPosition(tblSch); err != nil {
 		return nil, err
 	}
+	if collatedType, ok := a.column.Type.(sql.TypeWithCollation); ok && collatedType.Collation() == sql.Collation_Invalid {
+		a.column.Type = collatedType.WithNewCollation(alterable.Collation())
```

Comment:
Not sure about this. The table can have a collation, and the column can as well. A column having no collation is a valid thing that integrators have to encode. You can't pretend that the column actually has the table's collation in this case, that's very different semantically.

Reply:
I'm a bit confused on what you mean here. All columns that use collations are required to have a valid collation in order to be used. The concept of "no collation" doesn't exist outside of analysis. During the analysis phase (I'm counting the `RowIter` call as part of analysis), we're assigning the column a collation only if it didn't have a collation explicitly set, which is taken from the table's collation. It's not that the column _has_ the table's collation, but it's _assigned_ the table's collation. If the table's collation changes, it has no bearing on the column, as it has already been assigned a collation.

There's also the distinction that the column didn't have a collation explicitly set during parsing. I mentioned it in the last feedback comment, but having the invalid collation will fail if it makes it past this point. There's no other way to determine that the parser didn't explicitly set a collation otherwise, unless we add a boolean to the plan struct. That could then cause some code path to slip the wrong collation through though, as we'd default to a valid collation rather than an invalid one. I'd rather error if we miss something in analysis.

Reply:
Discussed offline.

MySQL's behavior on column / table collations is pretty confusing, please add a comment about this.